### PR TITLE
Add "name" / "value" support for Python `with`

### PR DESCRIPTION
**Pull request by pokey**
## Checklist

- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [-] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [-] I have not broken the cheatsheet

### CODE REVIEW

File: queries/python.scm [hunk 1]
```diff
@@ -142,6 +142,17 @@
...
 ) @_.domain
 
+;;!! with aaa:
+;;!       ^^^
+;;!  --------
+(with_statement
+  (with_clause
+    (with_item
+      value: (_) @value
+    )
```
Comment (auscompgeek) [inline queries/python.scm:152]: Do we want to add "name" too?

Comment (pokey) [reply queries/python.scm:152]: yeah was thinking that. presumably "value" shouldn't include the `as aaa` piece?

File: packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/python/value.resourceManagement5.scope
Comment (pokey) [inline packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/python/value.resourceManagement5.scope:1]: this is analogous to what we do for `const aaa = bbb, ccc = ddd;` in typescript, where there is one scope for each subclause and one scope for the entire statement

File: packages/common/src/scopeSupportFacets/scopeSupportFacets.types.ts [hunk 2]
```diff
@@ -76,6 +78,8 @@ const scopeSupportFacets = [
   "value.return",
   "value.return.lambda",
   "value.field",
+  "value.resourceManagement",
```
[the remark this answers is not in the document]
Comment (auscompgeek) [reply packages/common/src/scopeSupportFacets/scopeSupportFacets.types.ts:81]: Yeah, Java's try-with-resources was the first one to come to mind other than Python.

Don't forget the ECMAScript explicit resource management proposal

File: packages/common/src/scopeSupportFacets/scopeSupportFacets.types.ts [hunk 3]
```diff
@@ -63,6 +63,8 @@ const scopeSupportFacets = [
   "name.function",
   "name.class",
   "name.field",
+  "name.resourceManagement",
```
Comment (pokey) [inline packages/common/src/scopeSupportFacets/scopeSupportFacets.types.ts:66]: `resourceManagement` => `resource`

File: packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/python/name.resourceManagement2.scope [hunk 4]
```diff
@@ -0,0 +1,56 @@
+with aaa, bbb as ccc:
```
Comment (pokey) [inline packages/cursorless-vscode-e2e/src/suite/fixtures/scopes/python/name.resourceManagement2.scope:1]: make `aaa` a name as well

File: queries/python.scm [hunk 5]
```diff
@@ -142,6 +142,88 @@
...
+      )
+    )
+  ) @_.domain
+  (#allow-multiple! @value)
+  (#allow-multiple! @name)
+)
+
+;;!! with aaa, bbb:
+;;!       ^^^  ^^^
+;;!       ---  ---
```
Comment (pokey) [inline queries/python.scm:205]:
```suggestion
;;!! with aaa as ccc, bbb:
;;!       ^^^         ^^^
;;!       ----------  ---
```

File: queries/python.scm [hunk 6]
```diff
@@ -209,7 +291,16 @@
...
-(module) @statement.iteration @value.iteration @name.iteration
+(module) @statement.iteration
+
+;; This is a hack to handle the case where the entire document is a `with` statement
+(
+  (module
+    (_) @_statement
+  ) @value.iteration @name.iteration
+  (#not-type? @_statement "with_statement")
+)
```
Comment (pokey) [inline queries/python.scm:302]: make sure we still have iteration scope when there is a top-level with statement and other statments as well. eg:

```py
with aaa:
    pass

bbb = ccc
ddd = eee
```

Comment (pokey) [reply queries/python.scm:304]: confirmed

**Comment by pokey** [comment]
ok @AndreasArvidsson this one has been updated based on our discussion

**Comment by pokey** [comment]
Note that CI will fail until #2206 is merged

**Review by AndreasArvidsson — APPROVED** [review]

**Pull request merged**